File: playground/ggz-java5/src/ggz/common/XMLElement.java
/*
 * Copyright (C) 2006  Helg Bredow
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ggz.common;

import org.xml.sax.Attributes;

/**
 * Object representing a single XML element.
 * 
 * Except for "process", do not access these members directly. Instead use the
 * provided accessor functions. "process" is meant to be invoked as a method on
 * instances of GGZXMLElement.
 */
public class XMLElement {
    /** The name of the element. */
    private String tag;

    /** Text content of an element. */
    private StringBuffer text;

    /** List of attributes on the element. */
    private String[][] attributes;

    /** Extra data associated with tag (usually gleaned from children). */
    private Object data;

    public XMLElement(String tag, Attributes attrs) {
        this.tag = tag;
        if (attrs.getLength() > 0) {
            this.attributes = new String[attrs.getLength()][2];
            for (int i = 0; i < attrs.getLength(); i++) {
                this.attributes[i][0] = attrs.getQName(i);
                this.attributes[i][1] = attrs.getValue(i);
            }
        }
    }

    public void set_data(Object data) {
        this.data = data;
    }

    public String get_tag() {
        return this.tag;
    }

    public String get_attr(String attr) {
        if (this.attributes != null) {
            for (int i = 0; i < this.attributes.length; i++) {
                if (this.attributes[i][0].equals(attr)) {
                    return this.attributes[i][1];
                }
            }
        }
        return null;
    }

    public Object get_data() {
        return this.data;
    }

    public String get_text() {
        return this.text == null ? null : this.text.toString();
    }

    public void add_text(char[] ch, int start, int length) {
        if (text == null) {
            text = new StringBuffer();
        }
        text.append(ch, start, length);
    }
}

File: ggz/src/ggzcore/xmlelement.c
/*
 * File: xmlelement.c
 * Author: Brent Hendricks
 * Project: GGZ Core Client Lib
 * Date: 8/31/01
 *
 * This is the code for handling the an XML element
 *
 * Copyright (C) 2000,2001 Brent Hendricks.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <config.h>

#include <stdlib.h>
#include <string.h>

#include "ggzcore.h"
#include "xmlelement.h"
#include "msg.h"
#include "memory.h"

static void _ggzcore_xmlelement_do_free(GGZXMLElement *element);


GGZXMLElement* _ggzcore_xmlelement_new(char *tag, void (*process)(), void (*free)())
{
	GGZXMLElement *element;

	element = ggzcore_malloc(sizeof(GGZXMLElement));
	
	_ggzcore_xmlelement_init(element, tag, process, free);

	return element;
}


void _ggzcore_xmlelement_init(GGZXMLElement *element, char *tag, void (*process)(), void (*free)())
{
	if (element) {
		element->tag = ggzcore_strdup(tag);
		element->text = NULL;
		element->process = process;
		
		if (free)
			element->free = free;
		else
			element->free = _ggzcore_xmlelement_do_free;
	}
}


char* _ggzcore_xmlelement_get_tag(GGZXMLElement *element)
{
	return (element ? element->tag : NULL);
}


char* _ggzcore_xmlelement_get_text(GGZXMLElement *element)
{
	return (element ? element->text : NULL);
}


void _ggzcore_xmlelement_add_text(GGZXMLElement *element, const char *text, int len)
{
	int old_len = 0;
	int new_len = 0;

	if (element) {

		/* Allocate space for text if we haven't already */
		if (!element->text) {
			new_len = len + 1;
			element->text = ggzcore_malloc(new_len * sizeof(char));
			element->text[0] = '\0';
		}
		else {
			old_len = strlen(element->text);
			new_len = old_len + len + 1;
			element->text = ggzcore_realloc(element->text, new_len);
		}
		
		strncat(element->text, text, len);
		element->text[new_len - 1] = '\0';
	}
}


void _ggzcore_xmlelement_free(GGZXMLElement *element)
{
	if (element) {
		if (element->tag)
			ggzcore_free(element->tag);
		if (element->text)
			ggzcore_free(element->text);
		if (element->free)
			element->free(element);
	}
}


static void _ggzcore_xmlelement_do_free(GGZXMLElement *element)
{
	if (element) 
		ggzcore_free(element);
}
